compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -979,6 +979,17 @@ impl<'a> Parser<'a> {
                 // we are using noexpect here because we don't expect a `.` directly after a `return`
                 // which could be suggested otherwise
                 self.eat_noexpect(&token::Dot)
+            } else if self.token.kind == TokenKind::RArrow
+                && self.prev_token.span.overlaps_or_adjacent(self.token.span)
+            {
+                // Recovery for `expr->suffix`.
+                //
+                // We make sure the `expr` and `->` are adjacent because we don't want to trigger the
+                // recovery for function signature.
```

Comment:
Why would this happen?

Reply:
Removing this makes [issue-118530-ice.rs](https://github.com/rust-lang/rust/blob/c3ceb00281f9557dcf5bba54fc44c9931cc90d42/tests/ui/parser/issues/issue-118530-ice.rs#L5) fail. I tried hard to understand this test. My understanding is that line 5 is a function signature, but I'd love to be corrected.

While avoiding this failure by checking the spans is not ideal, this indeed works and is consistent with how people space their `->` in function prototypes and don't space their `->` when accessing fields.

Reply:
Also, i feel like removing the soacea around the -> in that test example should cause it to still ICE?

if so, this PR still introduces new bugs, and those should be fixed.

Reply:
> Removing this makes [issue-118530-ice.rs](https://github.com/rust-lang/rust/blob/c3ceb00281f9557dcf5bba54fc44c9931cc90d42/tests/ui/parser/issues/issue-118530-ice.rs#L5) fail.

I've only glanced at this test but it reminded me of something I wanted to write in my initial review: The snippet `Fn::() -> Option<i32>` is a syntactically valid expression, believe it or not. See https://github.com/rust-lang/rust/issues/120892#issuecomment-1937169035 for more context. I don't like that it's that way and we should probably change it but definitely not in this PR. Even if you check for “adjacency”, that won't suffice:

```rs
#[cfg(any())]
fn f() {
    let _ = Fn::()->Option<i32>;;
}
```

This should ideally continue to compile (in this PR).